src/hooks/useDebouncedSearch.ts

```diff
@@ -0,0 +1,74 @@
+import { MeiliSearch, SearchParams, SearchResponse } from "meilisearch";
+import { useCallback, useEffect, useMemo, useState } from "react";
+
+export const useDebouncedSearch = <T extends Record<string, any>>(
+  host: string,
+  apiKey: string,
+  indexName: string,
+  params: SearchParams,
+  debounceMs: number = 500,
+) => {
+  if (host === "") {
+    console.error(`useDebouncedSearch.host is missing`);
+  }
+  if (apiKey === "") {
+    console.error(`useDebouncedSearch.apiKey is missing`);
+  }
+  if (indexName === "") {
+    console.error(`useDebouncedSearch.indexName is missing`);
+  }
+
+  // Controlled input for rendering
+  const [rawInput, setRawInput] = useState("");
+  // Actual query from input that gets sent in search requests
+  const [query, setQuery] = useState("");
+
+  const [response, setResponse] = useState<SearchResponse<T> | null>(null);
+
+  // Get index
+  const index = useMemo(() => {
+    const meilisearch = new MeiliSearch({
+      host,
+      apiKey,
+    });
+    return meilisearch.index<T>(indexName);
+  }, [host, apiKey, indexName]);
+
+  // Get search response
+  const search = useCallback(async () => {
+    if (query === "") {
+      return;
+    }
+    try {
+      const response = await index.search<T>(query, params);
+      setResponse(response);
+    } catch (e) {
+      console.error(`Search for query "${query}" failed (${e})`);
+    }
+  }, [query]);
+
+  // Perform search
+  useEffect(() => {
```

Comment:
Suggestion: Use react-query here instead of `useEffect` as it will handle loading/error states better, as well as has a few QoL features for fetching data from APIs.